src/Parsers/ExpressionListParsers.cpp

```diff
@@ -599,158 +609,1974 @@ bool ParserLambdaExpression::parseImpl(Pos & pos, ASTPtr & node, Expected & expe
+
+    args_right_expr->children.emplace_back(arguments[0]);
+    args_right_expr->children.emplace_back(arguments[2]);
+
+    if (negative)
+    {
+        /// NOT BETWEEN
+        f_left_expr->name = "less";
+        f_right_expr->name = "greater";
+        f_combined_expression->name = "or";
+    }
+    else
+    {
+        /// BETWEEN
+        f_left_expr->name = "greaterOrEquals";
+        f_right_expr->name = "lessOrEquals";
+        f_combined_expression->name = "and";
+    }
+
+    f_left_expr->arguments = args_left_expr;
+    f_left_expr->children.emplace_back(f_left_expr->arguments);
+
+    f_right_expr->arguments = args_right_expr;
+    f_right_expr->children.emplace_back(f_right_expr->arguments);
+
+    args_combined_expression->children.emplace_back(f_left_expr);
+    args_combined_expression->children.emplace_back(f_right_expr);
+
+    f_combined_expression->arguments = args_combined_expression;
+    f_combined_expression->children.emplace_back(f_combined_expression->arguments);
+
+    return f_combined_expression;
 }
 
+enum Action
+{
+    OPERAND,
+    OPERATOR
+};
 
-bool ParserPrefixUnaryOperatorExpression::parseImpl(Pos & pos, ASTPtr & node, Expected & expected)
+enum OperatorType
 {
-    /// try to find any of the valid operators
-    const char ** it;
-    for (it = operators; *it; it += 2)
+    None,
+    Comparison,
+    Mergeable,
+    ArrayElement,
+    TupleElement,
+    IsNull,
+    StartBetween,
+    StartNotBetween,
+    FinishBetween,
+    StartIf,
+    FinishIf,
+    Cast
+};
+
+// class Operator:
+//   - defines structure of certain operator
+class Operator
+{
+public:
+    Operator() = default;
+
+    Operator(String function_name_,
+             Int32 priority_,
+             Int32 arity_ = 2,
+             OperatorType type_ = OperatorType::None) : type(type_), priority(priority_), arity(arity_), function_name(function_name_)
     {
-        if (parseOperator(pos, *it, expected))
-            break;
     }
 
-    /// Let's parse chains of the form `NOT NOT x`. This is hack.
-    /** This is done, because among the unary operators there is only a minus and NOT.
-      * But for a minus the chain of unary operators does not need to be supported.
-      */
-    size_t count = 1;
-    if (it[0] && 0 == strncmp(it[0], "NOT", 3))
+    OperatorType type;
+    Int32 priority;
+    Int32 arity;
+    String function_name;
+};
+
+class Layer
+{
+public:
+    virtual ~Layer() = default;
+
+    bool popOperator(Operator & op)
     {
-        while (true)
-        {
-            const char ** jt;
-            for (jt = operators; *jt; jt += 2)
-                if (parseOperator(pos, *jt, expected))
-                    break;
+        if (operators.empty())
+            return false;
 
-            if (!*jt)
-                break;
+        op = std::move(operators.back());
+        operators.pop_back();
 
-            ++count;
+        return true;
+    }
+
+    void pushOperator(Operator op, bool count = true)
```

Comment:
Here it is unusual to have a variable `count` which is `bool` but not `int`.
I don't understand what's the counting logic for depth here. Maybe add a comment?

Reply:
`count` was a part of some previous logic, so we don't need it anymore, I'll remove it :)